pkg/trace/stats/client_stats_aggregator.go

```diff
@@ -0,0 +1,301 @@
+		buckets:       make(map[int64]*bucket, 20),
+		out:           out,
+		agentEnv:      conf.DefaultEnv,
+		agentHostname: conf.Hostname,
+		oldestTs:      alignAgg(time.Now().Add(bucketDuration - oldestBucketStart)),
+		exit:          make(chan struct{}),
+		done:          make(chan struct{}),
+	}
+}
+
+// Start starts the aggregator.
+func (a *ClientStatsAggregator) Start() {
+	go func() {
+		defer watchdog.LogOnPanic()
+		for {
+			select {
+			case t := <-a.flushTicker.C:
+				a.flushOnTime(t)
+			case input := <-a.In:
+				a.add(time.Now(), input)
+			case <-a.exit:
+				a.flushAll()
+				close(a.done)
+				return
+			}
+		}
+	}()
+}
+
+// Stop stops the aggregator. Calling Stop twice will panic.
+func (a *ClientStatsAggregator) Stop() {
+	close(a.exit)
+	<-a.done
+}
+
+// flushOnTime flushes all buckets up to flushTs, except the last one.
+func (a *ClientStatsAggregator) flushOnTime(now time.Time) {
+	flushTs := alignAgg(now.Add(bucketDuration - oldestBucketStart))
+	for t := a.oldestTs; t.Before(flushTs); t = t.Add(bucketDuration) {
+		if b, ok := a.buckets[t.Unix()]; ok {
+			a.flush(b.flush())
+			delete(a.buckets, t.Unix())
+		}
+	}
+	a.oldestTs = flushTs
+}
+
+func (a *ClientStatsAggregator) flushAll() {
+	for _, b := range a.buckets {
+		a.flush(b.flush())
+	}
+}
+
+// getAggregationBucketTime returns unix time at which we aggregate the bucket.
+// We timeshift payloads older than a.oldestTs to a.oldestTs.
+// Payloads in the future are timeshifted to the latest bucket.
+func (a *ClientStatsAggregator) getAggregationBucketTime(now, bs time.Time) (time.Time, bool) {
+	if bs.Before(a.oldestTs) {
+		return a.oldestTs, true
+	}
+	if bs.After(now) {
+		return alignAgg(now), true
+	}
+	return alignAgg(bs), false
+}
+
+func (a *ClientStatsAggregator) add(now time.Time, p pb.ClientStatsPayload) {
+	for _, clientBucket := range p.Stats {
+		clientBucketStart := time.Unix(0, int64(clientBucket.Start))
+		ts, shifted := a.getAggregationBucketTime(now, clientBucketStart)
+		if shifted {
+			clientBucket.AgentTimeShift = ts.Sub(clientBucketStart).Nanoseconds()
+			clientBucket.Start = uint64(ts.UnixNano())
+		}
+		b, ok := a.buckets[ts.Unix()]
+		if !ok {
+			b = &bucket{ts: ts}
+			a.buckets[ts.Unix()] = b
+		}
+		p.Stats = []pb.ClientStatsBucket{clientBucket}
+		a.flush(b.add(p))
+	}
+}
+
+func (a *ClientStatsAggregator) flush(p []pb.ClientStatsPayload) {
+	if len(p) == 0 {
+		return
+	}
+	a.out <- pb.StatsPayload{
+		Stats:          p,
+		AgentEnv:       a.agentEnv,
+		AgentHostname:  a.agentHostname,
+		AgentVersion:   info.Version,
+		ClientComputed: true,
+	}
+}
+
+// alignAgg aligns time to the aggregator timestamps.
+// Timestamps from the aggregator are never aligned  with concentrator timestamps.
+// This ensures that all counts sent by a same agent host are never on the same second.
+// aggregator timestamps:   2ks+1s (1s, 3s, 5s, 7s, 9s, 11s)
+// concentrator timestamps: 10ks   (0s, 10s, 20s ..)
+func alignAgg(t time.Time) time.Time {
```

Comment:
```suggestion
func alignT(t time.Time) time.Time {
```

Reply:
I think this needs to be clearly distinguished from concentrator function alignTs, and I don't think alignT matches

Reply:
`alignTime` then. `alignAgg` is just a quick rename to not collide with the other one but "agg" has no place in the behaviour of this func. 

Or, if you really want to keep "agg", at least try `alignAggTs` or `alignAggTime`